packages/itwin/tree-widget/src/tree-widget-react/components/trees/categories-tree/internal/visibility/CategoriesTreeVisibilityHelper.ts

```diff
@@ -0,0 +1,129 @@
+    return from(this.#props.idsCache.getAllContainedCategories(props.definitionContainerIds)).pipe(
+      mergeMap((categoryIds) =>
+        this.getCategoriesVisibilityStatus({
+          categoryIds,
+          modelId: undefined,
+          type: this.#props.viewport.view.is2d() ? "DrawingCategory" : "SpatialCategory",
+        }),
+      ),
+    );
+  }
+
+  /** Gets grouped elements visibility status. */
+  public getGroupedElementsVisibilityStatus(props: { modelElementsMap: Map<ModelId, Set<ElementId>>; categoryId: Id64String }): Observable<VisibilityStatus> {
+    const { modelElementsMap, categoryId } = props;
+    return from(modelElementsMap).pipe(
+      mergeMap(([modelId, elementIds]) =>
+        this.getElementsVisibilityStatus({
+          elementIds,
+          modelId,
+          categoryId,
+          type: this.#props.viewport.view.is2d() ? "GeometricElement2d" : "GeometricElement3d",
+        }),
+      ),
+      mergeVisibilityStatuses,
+    );
+  }
+
+  /**
+   * Changes visibility status of definition containers.
+   *
+   * Does this by changing visibility status of related categories.
+   */
+  public changeDefinitionContainersVisibilityStatus(props: { definitionContainerIds: Id64Arg; on: boolean }): Observable<void> {
+    return from(this.#props.idsCache.getAllContainedCategories(props.definitionContainerIds)).pipe(
+      mergeMap((categoryIds) => this.changeCategoriesVisibilityStatus({ categoryIds, modelId: undefined, on: props.on })),
+    );
+  }
+
+  /**
+   * Changes visibility status of sub-categories.
+   *
+   * Also, enables parent categories if `on` is true.
+   */
+  public changeSubCategoriesVisibilityStatus(props: { categoryId: Id64String; subCategoryIds: Id64Arg; on: boolean }): Observable<void> {
+    return concat(
+      // make sure parent category and models are enabled
+      props.on
+        ? concat(
+            from(enableCategoryDisplay(this.#props.viewport, props.categoryId, props.on, false)),
+            this.enableCategoriesElementModelsVisibility(props.categoryId),
+          )
+        : EMPTY,
+      from(props.subCategoryIds).pipe(map((subCategoryId) => enableSubCategoryDisplay(this.#props.viewport, subCategoryId, props.on))),
+    );
+  }
+
+  /** Changes grouped elements visibility status. */
+  public changeGroupedElementsVisibilityStatus(props: {
+    modelElementsMap: Map<ModelId, Set<ElementId>>;
+    categoryId: Id64String;
+    on: boolean;
+  }): Observable<void> {
+    return from(props.modelElementsMap).pipe(
+      mergeMap(([modelId, elementIds]) => {
+        return this.changeElementsVisibilityStatus({ modelId, elementIds, categoryId: props.categoryId, on: props.on });
+      }),
+    );
+  }
+
+  /** Turns on visiblity status of models (that are not yet turned on) that are related to categories. */
+  private enableCategoriesElementModelsVisibility(categoryIds: Id64Arg): Observable<void> {
```

Comment:
any reason for this method to be named differently than others ("enable" vs "change", "visibility" vs "visibility status")?